train.py

```diff
@@ -352,11 +368,72 @@ def get_optimizer(model:nn.Module, args:Dict):
     elif args["optimizer"] == "adadelta":
         return optim.Adadelta(model.parameters(), lr=args['lr'])
     elif args["optimizer"] == "adamw":
-        return torch.optim.AdamW(model.parameters(), lr=args['lr'], betas=(0.9,0.95), eps=1e-5)
+        return torch.optim.AdamW(model.parameters(), lr=args['lr'], betas=(0.9,0.95),
+                                 eps=1e-5, weight_decay=args['wd'])
     else:
         raise ValueError("Invalid optimizer")
 
 
+# Wrap the model (LoRA policy from llama-recipes):
+# This checks for lora layers (has weight and requires_grad)
+def create_default_auto_wrap_policy():
+    def lambda_policy_fn(module):
+        return (
+            len(list(module.named_children())) == 0
+            and getattr(module, "weight", None) is not None
+            and module.weight.requires_grad
+        )
+    lambda_policy = functools.partial(lambda_auto_wrap_policy, lambda_fn=lambda_policy_fn)
+    transformer_layer_name = LlamaDecoderLayer
+    transformer_wrap_policy = functools.partial(
+        transformer_auto_wrap_policy,
+        transformer_layer_cls=(
+            PrefixEncoder,
+            PromptEncoder,
+            PromptEmbedding,
+            transformer_layer_name,
+        ),
+    )
+    return functools.partial(_or_policy, policies=[lambda_policy, transformer_wrap_policy])
+
+
+# Custom QLORA module.
+class QLORA(nn.Module):
+    def __init__(self, base_layer, lora_rank, lora_alpha, lora_dropout, device="cpu"):
```

Comment:
Would it be better to init the LoRA layers on the `base_layer.device` rather than CPU?

Reply:
Does it matter -- they're really small anyway right?

Reply:
That was also my thinking also, but made the change nevertheless.